main.cc

```diff
@@ -1101,12 +1101,28 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // described here: https://github.com/scylladb/scylla/issues/1014
             supervisor::notify("loading system sstables");
             replica::distributed_loader::init_system_keyspace(sys_ks, erm_factory, db, *cfg, system_table_load_phase::phase1).get();
-            cfg->host_id = sys_ks.local().load_local_host_id().get0();
+
+            auto listen_address = utils::resolve(cfg->listen_address, family).get0();
+
+            auto linfo = sys_ks.local().load_local_info().get0();
+            if (linfo.cluster_name.empty()) {
+                linfo.cluster_name = cfg->cluster_name();
+            } else if (linfo.cluster_name != cfg->cluster_name()) {
+                throw exceptions::configuration_exception("Saved cluster name " + linfo.cluster_name + " != configured name " + cfg->cluster_name());
+            }
+            if (!linfo.host_id) {
+                linfo.host_id = locator::host_id::create_random_id();
```

Comment:
Both host_id and cluster_name have reasonable emptyness checks. Not a huge fan (those empty values can also be set in bad config), but probably okay.